airbyte-integrations/connectors/source-mssql/src/main/java/io/airbyte/integrations/source/mssql/initialsync/MssqlInitialReadUtil.java

```diff
@@ -0,0 +1,332 @@
+                                   Map<AirbyteStreamNameNamespacePair, OrderedColumnLoadStatus> pairToInitialLoadStatus) {
+
+  }
+
+  public record CursorBasedStreams(List<ConfiguredAirbyteStream> streamsForCursorBased,
+                                   Map<AirbyteStreamNameNamespacePair, CursorBasedStatus> pairToCursorBasedStatus) {
+
+  }
+
+  public record OrderedColumnInfo(String ocFieldName, JDBCType fieldType, String ocMaxValue) {}
+
+  public static List<AutoCloseableIterator<AirbyteMessage>> getCdcReadIterators(final JdbcDatabase database,
+                                                                                final ConfiguredAirbyteCatalog catalog,
+                                                                                final Map<String, TableInfo<CommonField<JDBCType>>> tableNameToTable,
+                                                                                final StateManager stateManager,
+                                                                                final Instant emittedAt,
+                                                                                final String quoteString) {
+    final JsonNode sourceConfig = database.getSourceConfig();
+    final Duration firstRecordWaitTime = RecordWaitTimeUtil.getFirstRecordWaitTime(sourceConfig);
+    final Duration subsequentRecordWaitTime = RecordWaitTimeUtil.getSubsequentRecordWaitTime(sourceConfig);
+    LOGGER.info("First record waiting time: {} seconds", firstRecordWaitTime.getSeconds());
+    // Determine the streams that need to be loaded via primary key sync.
+    final List<AutoCloseableIterator<AirbyteMessage>> initialLoadIterator = new ArrayList<>();
+    // Construct the initial state for Mssql. If there is already existing state, we use that instead
+    // since that is associated with the debezium state associated with the initial sync.
+    final MssqlDebeziumStateUtil mssqlDebeziumStateUtil = new MssqlDebeziumStateUtil();
+    final JsonNode initialDebeziumState = mssqlDebeziumStateUtil.constructInitialDebeziumState(
+        getDebeziumProperties(database, catalog, false), catalog, database);
+
+    final JsonNode state =
+        (stateManager.getCdcStateManager().getCdcState() == null || stateManager.getCdcStateManager().getCdcState().getState() == null)
+            ? initialDebeziumState
+            : Jsons.clone(stateManager.getCdcStateManager().getCdcState().getState());
+
+    final Optional<MssqlDebeziumStateAttributes> savedOffset = mssqlDebeziumStateUtil.savedOffset(
+        getDebeziumProperties(database, catalog, true), catalog, state.get(MSSQL_CDC_OFFSET), sourceConfig);
+    final boolean savedOffsetStillPresentOnServer =
+        savedOffset.isPresent() && mssqlDebeziumStateUtil.savedOffsetStillPresentOnServer(database, savedOffset.get());
+
+    if (!savedOffsetStillPresentOnServer) {
+      LOGGER.warn("Saved offset no longer present on the server, Airbyte is going to trigger a sync from scratch");
```

Comment:
can we just throw an exception if the saved offset is no longer present? This is line with new behavior telling the user to reset their streams

Reply:
I'll take this offline - need to think since user will see more failures as opposed to connector handling this case quietly.
To compare: postgres and mysql are still going to fall back to initial and not throw an error.